lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,969 @@
+      std::is_same<A, parser::EndDoStmt>,
+      std::is_same<A, common::Indirection<parser::EndDoStmt>>,
+      std::is_same<A, parser::BlockStmt>, std::is_same<A, parser::EndBlockStmt>,
+      std::is_same<A, parser::CriticalStmt>,
+      std::is_same<A, parser::EndCriticalStmt>,
+      std::is_same<A, parser::ForallConstructStmt>,
+      std::is_same<A, parser::ForallStmt>,
+      std::is_same<A, parser::WhereConstructStmt>,
+      std::is_same<A, parser::EndFunctionStmt>,
+      std::is_same<A, parser::EndMpSubprogramStmt>,
+      std::is_same<A, parser::EndProgramStmt>,
+      std::is_same<A, parser::EndSubroutineStmt>>;
+}
+template<>
+constexpr bool IsLegalBranchTarget(
+    const parser::Statement<parser::ActionStmt> &actionStmt) {
+  if (!isStrictF18) {
+    return true;
+  } else {
+    // XXX: do we care to flag these as errors? If we want strict F18, these
+    // statements should not even be present
+    return !(
+        std::holds_alternative<common::Indirection<parser::ArithmeticIfStmt>>(
+            actionStmt.statement.u) ||
+        std::holds_alternative<common::Indirection<parser::AssignStmt>>(
+            actionStmt.statement.u) ||
+        std::holds_alternative<common::Indirection<parser::AssignedGotoStmt>>(
+            actionStmt.statement.u) ||
+        std::holds_alternative<common::Indirection<parser::PauseStmt>>(
+            actionStmt.statement.u));
+  }
+}
+
+template<typename A>
+constexpr TargetStmtType ConsTrgtFlags(const parser::Statement<A> &statement) {
+  TargetStmtType targetStmtType{};
+  if (IsLegalDoTerm(statement)) {
+    targetStmtType.set(TargetStatementEnum::Do);
+  }
+  if (IsLegalBranchTarget(statement)) {
+    targetStmtType.set(TargetStatementEnum::Branch);
+  }
+  if (IsFormat(statement)) {
+    targetStmtType.set(TargetStatementEnum::Format);
+  }
+  return targetStmtType;
+}
+
+/// \brief \p opt1 and \p opt2 must be either present and identical or absent
+/// \param a  an optional construct-name (opening statement)
+/// \param b  an optional construct-name (ending statement)
+template<typename A> inline bool BothEqOrNone(const A &a, const A &b) {
+  return (a.has_value() == b.has_value())
+      ? (a.has_value() ? (a.value().ToString() == b.value().ToString()) : true)
+      : false;
+}
+
+/// \brief \p opt1 must either be absent or identical to \p b
+/// \param a  an optional construct-name for an optional constraint
+/// \param b  an optional construct-name (opening statement)
+template<typename A> inline bool PresentAndEq(const A &a, const A &b) {
+  return (!a.has_value()) ||
+      (b.has_value() && (a.value().ToString() == b.value().ToString()));
+}
+
+/// \brief Iterates over parse tree, creates the analysis result
+/// As a side-effect checks the constraints for the usages of
+/// <i>construct-name</i>.
+struct ParseTreeAnalyzer {
+public:
+  struct UnitAnalysis {
+  public:
+    SourceStmtList doStmtSources_;  ///< bases of label-do-stmts
```

Comment:
If class data members are public, use a `struct` and don't affix `_` to their names.  But for most cases apart from simple tuple-like containers, use `class`, make the data members private, and retain the `_`.

Reply:
_ added and _ deleted